File: skills/mobius-concept-map/scripts/build_mobius.py
#!/usr/bin/env python3
"""
莫比乌斯环概念地图生成脚本 v2.0
用法:
  python3 build_mobius.py --data data.json --title "标题" --output 输出.html
  python3 build_mobius.py --data data.json --no-emptiness --output 输出.html
  python3 build_mobius.py --data data.json --emptiness '{"wane":false}' --output 输出.html

注：生成的 HTML 内置五阶音景（Web Audio API）和动画录制功能（桌面端MP4/移动端WebM），需通过 HTTP 服务器访问。
"""
import json, argparse, subprocess, re
from pathlib import Path

TEMPLATE = Path(__file__).parent.parent / 'assets' / 'template.html'

# 默认五空配置
DEFAULT_EMPTINESS = {
    'sunyata': True, 'mirror': True, 'breath': True, 'wane': True, 'ascend': True,
    'mirrorRatio': 0.42, 'breathPeriod': 10000, 'breathAmplitude': 15,
    'fadeStart': 35000, 'fadeEnd': 50000, 'fadeTarget': 0.15, 'fadeRestore': 0.5,
    'ascendRatio': 0.2
}

def esc_js(s):
    """转义字符串用于 JS 单引号"""
    return s.replace('\\', '\\\\').replace("'", "\\'").replace('\n', '\\n')

def concept_to_js(c):
    """单个概念 → JS对象字符串"""
    docs_str = json.dumps(c.get('docs', []), ensure_ascii=False)
    line = "{{ label: '{l}', desc: '{d}', docs: {docs} }}".format(
        l=esc_js(c['label']), d=esc_js(c.get('desc', '')), docs=docs_str)
    # 注入可选字段
    if c.get('wuxing'):
        line = line[:-2] + ", wuxing:'" + c['wuxing'] + "' }"
    if c.get('bagua'):
        line = line[:-2] + ", bagua:'" + c['bagua'] + "' }"
    return line

def build_rings_js(rings):
    """整个rings数据 → JS字符串"""
    lines = ['var DEFAULT_RINGS = [']
    for ri, r in enumerate(rings):
        lines.append('    {')
        lines.append("      label: '{l}', R: {R}, w: {w}, yOffset: {yO}, color: '{c}',".format(
            l=esc_js(r['label']),
            R=r.get('R', 150), w=r.get('w', 25), yO=r.get('yOffset', 0),
            c=r.get('color', '#4ECDC4')))
        lines.append('      concepts: [')
        for c in r.get('concepts', []):
            lines.append('        ' + concept_to_js(c) + ',')
        lines.append('      ]')
        lines.append('    }' + (',' if ri < len(rings) - 1 else ''))
    lines.append('  ];')
    return '\n'.join(lines)

def build_emptiness_js(emp):
    """EMPTINESS 配置 → JS字符串"""
    lines = ['  var EMPTINESS = {']
    items = list(emp.items())
    for i, (k, v) in enumerate(items):
        if isinstance(v, bool):
            lines.append('    {}: {}'.format(k, 'true' if v else 'false'))
        elif isinstance(v, (int, float)):
            lines.append('    {}: {}'.format(k, v))
        else:
            lines.append("    {}: '{}'".format(k, v))
        if i < len(items) - 1:
            lines[-1] += ','
    lines.append('  };')
    return '\n'.join(lines)

def main():
    parser = argparse.ArgumentParser(description='莫比乌斯环概念地图生成器 v2.0')
    parser.add_argument('--data', required=True, help='概念数据 JSON 文件')
    parser.add_argument('--title', default='概念地图', help='页面标题')
    parser.add_argument('--output', default='output.html', help='输出 HTML 文件')
    parser.add_argument('--no-emptiness', action='store_true', help='关闭全部五空动画')
    parser.add_argument('--emptiness', default=None, help='自定义 emptiness JSON 字符串')
    args = parser.parse_args()

    with open(args.data, encoding='utf-8') as f:
        data = json.load(f)
    rings = data.get('rings', data if isinstance(data, list) else [])

    with open(TEMPLATE, encoding='utf-8') as f:
        html = f.read()

    # 替换 DEFAULT_RINGS
    old = 'var DEFAULT_RINGS = ['
    old_end = '  ];'
    si = html.index(old)
    ei = html.index(old_end, si) + len(old_end)
    html = html.replace(html[si:ei], build_rings_js(rings))

    # 构建 emptiness 配置
    if args.no_emptiness:
        emp = {k: False for k, v in DEFAULT_EMPTINESS.items() if isinstance(v, bool)}
    elif args.emptiness:
        emp = {**DEFAULT_EMPTINESS, **json.loads(args.emptiness)}
    elif isinstance(data, dict) and 'emptiness' in data:
        emp = {**DEFAULT_EMPTINESS, **data['emptiness']}
    else:
        emp = DEFAULT_EMPTINESS

    # 替换 EMPTINESS 配置
    emp_old = '  var EMPTINESS = {'
    # 找到 EMPTINESS 对象的开始和结束
    emp_si = html.index(emp_old)
    # 找到匹配的 }; 结束
    brace_count = 0
    emp_ei = emp_si
    for i in range(emp_si, len(html)):
        if html[i] == '{':
            brace_count += 1
        elif html[i] == '}':
            brace_count -= 1
            if brace_count == 0:
                emp_ei = i + 2  # 包含 };\n
                break
    html = html.replace(html[emp_si:emp_ei], build_emptiness_js(emp))

    # 替换标题
    html = html.replace('唯识论 · 种子现行 · v1.2 修正版', args.title)
    html = html.replace('唯识论 · 种子现行 · 概念地图', args.title)
    html = html.replace('莫比乌斯环概念地图 · v2.0', args.title)

    # 替换 Legend 三层标注
    legend_lines = []
    for r in rings:
        label = r.get('label', '未命名')
        color = r.get('color', '#4ECDC4')
        first_concept = r.get('concepts', [{}])[0].get('label', '')
        desc = ' · ' + first_concept if first_concept else ''
        legend_lines.append(
            '<div><span class="dot" style="background:{}"></span>{}{}</div>'.format(color, label, desc))
    html = html.replace('<!--LEGEND_RINGS-->', '\n'.join(legend_lines))

    with open(args.output, 'w', encoding='utf-8') as f:
        f.write(html)

    # 验证 JS 语法
    import shlex
    abs_output = str(Path(args.output).resolve())
    r = subprocess.run(['node', '-e',
        "const m=require('fs').readFileSync(process.argv[1],'utf8').match(/<script>([\\s\\S]*?)<\\/script>/);new Function(m[1]);console.log('OK')",
        abs_output],
        capture_output=True, text=True)
    ok = 'OK' in r.stdout
    total_nodes = sum(len(rn.get('concepts', [])) for rn in rings)
    emp_status = '五空全开' if all(v for k, v in emp.items() if isinstance(v, bool)) else '部分关闭'
    print("JS: {} | 节点: {} | 层: {} | {} | {}".format(
        '✅' if ok else '❌', total_nodes, len(rings), emp_status, args.output))
    if not ok:
        print(r.stderr[:500])

if __name__ == '__main__':
    main()

File: meta-skill-org/build_mobius.js
/**
 * 莫比乌斯环概念地图构建引擎 v2.0 (JS)
 * 从 Python build_mobius.py 移植
 */
(function () {
  'use strict';

  // 默认五空配置
  var DEFAULT_EMPTINESS = {
    sunyata: true, mirror: true, breath: true, wane: true, ascend: true,
    mirrorRatio: 0.42, breathPeriod: 10000, breathAmplitude: 15,
    fadeStart: 35000, fadeEnd: 50000, fadeTarget: 0.15, fadeRestore: 0.5,
    ascendRatio: 0.2
  };

  /**
   * 转义字符串用于 JS 单引号
   */
  function escJs(s) {
    return s.replace(/\\/g, '\\\\').replace(/'/g, "\\'").replace(/\n/g, '\\n');
  }

  /**
   * 单个概念 → JS 对象字符串
   */
  function conceptToJs(c) {
    var docsStr = JSON.stringify(c.docs || [], undefined, 0);
    var line = "{ label: '" + escJs(c.label) + "', desc: '" + escJs(c.desc || '') + "', docs: " + docsStr + " }";
    if (c.wuxing) {
      line = line.slice(0, -2) + ", wuxing:'" + escJs(c.wuxing) + "' }";
    }
    if (c.bagua) {
      line = line.slice(0, -2) + ", bagua:'" + escJs(c.bagua) + "' }";
    }
    return line;
  }

  /**
   * 整个 rings 数据 → JS 字符串
   */
  function buildRingsJs(rings) {
    var lines = ['var DEFAULT_RINGS = ['];
    for (var ri = 0; ri < rings.length; ri++) {
      var r = rings[ri];
      lines.push('    {');
      lines.push("      label: '" + escJs(r.label) + "', R: " + (r.R || 150) + ", w: " + (r.w || 25) + ", yOffset: " + (r.yOffset || 0) + ", color: '" + (r.color || '#4ECDC4') + "',");
      lines.push('      concepts: [');
      var concepts = r.concepts || [];
      for (var ci = 0; ci < concepts.length; ci++) {
        lines.push('        ' + conceptToJs(concepts[ci]) + ',');
      }
      lines.push('      ]');
      lines.push('    }' + (ri < rings.length - 1 ? ',' : ''));
    }
    lines.push('  ];');
    return lines.join('\n');
  }

  /**
   * EMPTINESS 配置 → JS 字符串
   */
  function buildEmptinessJs(emp) {
    var lines = ['  var EMPTINESS = {'];
    var keys = Object.keys(emp);
    for (var i = 0; i < keys.length; i++) {
      var k = keys[i];
      var v = emp[k];
      if (typeof v === 'boolean') {
        lines.push('    ' + k + ': ' + (v ? 'true' : 'false'));
      } else if (typeof v === 'number') {
        lines.push('    ' + k + ': ' + v);
      } else {
        lines.push("    " + k + ": '" + escJs(String(v)) + "'");
      }
      if (i < keys.length - 1) {
        lines[lines.length - 1] += ',';
      }
    }
    lines.push('  };');
    return lines.join('\n');
  }

  /**
   * djb2 字符串哈希函数
   */
  function djb2(str) {
    var hash = 5381;
    for (var i = 0; i < str.length; i++) {
      hash = ((hash << 5) + hash) + str.charCodeAt(i);
    }
    return Math.abs(hash).toString(16).substring(0, 8);
  }

  /**
   * 生成 localStorage 的 key
   */
  function generateStorageKey(title) {
    return 'mobius_' + djb2(title);
  }

  /**
   * 主构建函数
   * @param {string} templateHtml - template.html 的完整内容
   * @param {object} data - JSON 数据（包含 rings 数组）
   * @param {string} title - 页面标题
   * @param {object} emptinessConfig - 可选，自定义 emptiness 配置
   * @returns {string} 完整的 HTML 字符串
   */
  function build(templateHtml, data, title, emptinessConfig) {
    title = title || '概念地图';
    var html = templateHtml;

    // 提取 rings
    var rings = data.rings || (Array.isArray(data) ? data : []);

    // 1. 替换 DEFAULT_RINGS
    var oldStart = 'var DEFAULT_RINGS = [';
    var oldEnd = '  ];';
    var si = html.indexOf(oldStart);
    var ei = html.indexOf(oldEnd, si) + oldEnd.length;
    html = html.substring(0, si) + buildRingsJs(rings) + html.substring(ei);

    // 2. 构建 emptiness 配置
    var emp;
    if (emptinessConfig) {
      // 使用用户自定义配置（合并默认值）
      emp = {};
      for (var k in DEFAULT_EMPTINESS) {
        if (DEFAULT_EMPTINESS.hasOwnProperty(k)) {
          emp[k] = DEFAULT_EMPTINESS[k];
        }
      }
      if (typeof emptinessConfig === 'object') {
        for (var ck in emptinessConfig) {
          if (emptinessConfig.hasOwnProperty(ck)) {
            emp[ck] = emptinessConfig[ck];
          }
        }
      }
    } else if (typeof data === 'object' && !Array.isArray(data) && data.emptiness) {
      emp = {};
      for (var dk in DEFAULT_EMPTINESS) {
        if (DEFAULT_EMPTINESS.hasOwnProperty(dk)) {
          emp[dk] = DEFAULT_EMPTINESS[dk];
        }
      }
      for (var ek in data.emptiness) {
        if (data.emptiness.hasOwnProperty(ek)) {
          emp[ek] = data.emptiness[ek];
        }
      }
    } else {
      emp = {};
      for (var fk in DEFAULT_EMPTINESS) {
        if (DEFAULT_EMPTINESS.hasOwnProperty(fk)) {
          emp[fk] = DEFAULT_EMPTINESS[fk];
        }
      }
    }

    // 替换 EMPTINESS 配置
    var empOld = '  var EMPTINESS = {';
    var empSi = html.indexOf(empOld);
    var braceCount = 0;
    var empEi = empSi;
    for (var j = empSi; j < html.length; j++) {
      if (html[j] === '{') braceCount++;
      else if (html[j] === '}') {
        braceCount--;
        if (braceCount === 0) {
          empEi = j + 2; // 包含 };\n
          break;
        }
      }
    }
    html = html.substring(0, empSi) + buildEmptinessJs(emp) + html.substring(empEi);

    // 3. 替换三个标题
    html = html.replace('唯识论 \u00b7 种子现行 \u00b7 v1.2 修正版', title);
    html = html.replace('唯识论 \u00b7 种子现行 \u00b7 概念地图', title);
    html = html.replace('莫比乌斯环概念地图 \u00b7 v2.0', title);

    // 4. 替换 Legend 三层标注
    var legendLines = [];
    for (var li = 0; li < rings.length; li++) {
      var ring = rings[li];
      var label = ring.label || '未命名';
      var color = ring.color || '#4ECDC4';
      var firstConcept = (ring.concepts && ring.concepts[0]) ? ring.concepts[0].label : '';
      var desc = firstConcept ? ' \u00b7 ' + firstConcept : '';
      legendLines.push('<div><span class="dot" style="background:' + color + '"></span>' + label + desc + '</div>');
    }
    html = html.replace('<!--LEGEND_RINGS-->', legendLines.join('\n'));

    // 5. 替换 STORAGE_KEY
    var storageKey = generateStorageKey(title);
    html = html.replace("var STORAGE_KEY = 'weishi_mobius_data'", "var STORAGE_KEY = '" + storageKey + "'");

    return html;
  }

  // 导出为全局对象
  window.MobiusBuilder = {
    build: build,
    generateStorageKey: generateStorageKey,
    DEFAULT_EMPTINESS: DEFAULT_EMPTINESS
  };

})();
